Create a Flatpak Package For GTK App

BrainBlasted opened the pull request:
This creates a development flatpak package for the desktop app. Includes a rework of the non-elementary version's app ID and a fix for an issue introduced in #28

BrainBlasted committed:
Rename App ID to fit within freedesktop guidelines

Previously the app used "writeas-gtk" as the app id. This was invalid
according to freedesktop specifications.

See https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-id-generic
for details on what an App ID should be.

BrainBlasted committed:
flatpak: Create flatpak manifest

Creates a flatpak manifest to build the app.
Does not yet build the writeas-cli.

Related to T663

BrainBlasted committed:
gitignore: Add common build directories from flatpak

These are the directories flatpak can create while building
an app. Since they should not be committed, they belong
in the gitignore.

BrainBlasted committed:
build: Rework cli script setup for flatpak

Flatpak doesn't support running extra commands, so we need
to make the cli build it's own target. This is made optional
since the cli doesn't always need to be shipped with the app.

Related to T663

BrainBlasted committed:
build: Properly set executable name in all files

This was missed during the consolidation of repositiories.
Since the app has different executable names for the
different builds, the app_id cannot be used as the "exec"
line unconditionally.

BrainBlasted committed:
debian: Remove "build" run_target usage

This run_target no longer exists.

BrainBlasted commented:
The current issues:

* ~~There's no way to authenticate from the UI, which means you can't actually publish from the app~~
* [x] The current way of saving drafts doesn't work with the flatpak sandbox. I'll look into that - either we make a sandbox hole or I figure this out.
* [x] It requires xclip currently. I'll need to see if there's a way around this or if I need to ship xclip within the flatpak.

BrainBlasted commented:
Once we get the last two worked out can we merge this? It will be much easier for me to work on number 1 after that

BrainBlasted commented:
In addition: should I look into getting flatpak CI set up, or should I make that a separate task?

thebaer commented:
> * There's no way to authenticate from the UI, which means you can't actually publish from the app

Hm, you should be able to publish -- the `writeas-cli` doesn't require authentication to publish. Maybe it's related to the flatpak sandbox? Because the CLI does need access to `~/.writeas`

> In addition: should I look into getting flatpak CI set up, or should I make that a separate task?

I'd say we address this as a separate task. Please feel free to create that in Phabricator

BrainBlasted commented:
Ah, apologies. The xclip issue is the only one that blocks publishing, then.

BrainBlasted committed:
flatpak: Add clipboard tooling

Fixes publishing and copies the post link to the clipboard.

BrainBlasted commented:
Fixed the publishing issue, will look into the draft issue next.

BrainBlasted committed:
flatpak: Add sandbox hole for ~/.writeas

Allows for flatpak users to have persistent drafts,
and for the bundled cli to access them.

BrainBlasted commented:
For now I added a sandbox hole to the flatpak. In the future I think we should migrate to using xdg dirs instead of a directory in ~/. Flatpak properly handles that use case by default.

BrainBlasted committed:
window: Use FileChooserNative instead of FileChooserDialog

Required for saving to work within flatpaks. Does not affect
non-flatpak usage.

Related to T663

BrainBlasted commented:
I'd say that this is ready now.

thebaer commented:
Great, sounds good. How can I test this out?

BrainBlasted commented:
There are a few ways: 

* [GNOME Builder](https://wiki.gnome.org/Apps/Builder) **(recommended)**. This should get you set up in a couple clicks. 

* Commands:

```
$ flatpak-builder --repo=repo app flatpak/com.abunchtell.WriteAsDesktop.json
```

Or I can send you a flatpak bundle. Which way would you prefer?

thebaer commented:
Thanks! I'd like to personally know how everything works, so I'm happy to build everything

BrainBlasted committed:
flatpak: Ensure read-write access to .writeas dir

Needed to make sure we can create, read, and write to this
directory.

ghost commented:
I'm having a tough time getting this working, now it's an error when running the built flatpak with `flatpak run com.abunchtell.WriteAsDesktop`:

```
Failed to register: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: org.freedesktop.DBus.Error.ServiceUnknown
```

This was after adding the local repo and installing for my user.

@thebaer needed dependencies with flatpak:
- `flatpak install org.gnome.Sdk//3.32`
- `flatpak install org.gnome.Platform//3.32`
- `flatpak install io.elementary.BaseApp//Juno`
- `flatpak install org.freedesktop.Sdk.Extension.golang//18.08`

@BrainBlasted I couldn't find an easy way to install required dependencies so just kept installing whatever the builder comlained was missing.

BrainBlasted commented:
@robjloranger do you have xdg-desktop-portal & xdg-desktop-portal-gtk installed?

ghost commented:
@BrainBlasted tried again today and found another error building:

```
[0/1] Installing files.
Installing src/wl-copy to /app/bin
Installing src/wl-paste to /app/bin
Installing /run/build/wl-clipboard/data/wl-copy.1 to /app/share/man/man1
Installing /run/build/wl-clipboard/data/wl-paste.1 to /app/share/man/man1
Installing /run/build/wl-clipboard/data/wl-clipboard.1 to /app/share/man/man1
Installing /run/build/wl-clipboard/completions/bash/wl-copy to /app/share/bash-completion/completions
Installing /run/build/wl-clipboard/completions/bash/wl-paste to /app/share/bash-completion/completions
Installing /run/build/wl-clipboard/completions/zsh/_wl-copy to /app/share/zsh/site-functions
Installing /run/build/wl-clipboard/completions/zsh/_wl-paste to /app/share/zsh/site-functions
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/mesonbuild/mesonmain.py", line 122, in run
    return options.run_func(options)
  File "/usr/lib/python3.7/site-packages/mesonbuild/minstall.py", line 514, in run
    installer.do_install(datafilename)
  File "/usr/lib/python3.7/site-packages/mesonbuild/minstall.py", line 346, in do_install
    self.install_data(d)
  File "/usr/lib/python3.7/site-packages/mesonbuild/minstall.py", line 375, in install_data
    d.dirmaker.makedirs(outdir, exist_ok=True)
  File "/usr/lib/python3.7/site-packages/mesonbuild/minstall.py", line 55, in makedirs
    os.makedirs(path, exist_ok=exist_ok)
  File "/usr/lib/python3.7/os.py", line 211, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/usr/lib/python3.7/os.py", line 221, in makedirs
    mkdir(name, mode)
OSError: [Errno 30] Read-only file system: '/usr/share/fish'
FAILED: meson-install 
/usr/bin/meson install --no-rebuild
ninja: build stopped: subcommand failed.
Error: module wl-clipboard: Child process exited with code 1
```

What version of python is required to build?

Also, are there any python modules required to build?

BrainBlasted commented:
This isn't a python issue by rather an idea where a dependency requires a new cli flag. Working on updating things.

BrainBlasted committed:
Build CLI with Go modules

BrainBlasted committed:
data: Validate appdata

Adds a test to validate appdata and fixes a mistake
that caused validation failure
